Use pundit policy_namespace in controllers

bhtabor opened the pull request:
Add pundit policy_namespace to controllers
Deprecate resolve_admin from pundit policies

bhtabor commented:
Fixes https://github.com/thoughtbot/administrate/issues/1332

To use specific [Policy Namespacing](https://github.com/varvet/pundit#policy-namespacing) for administrate, like admin namespaced policies for example, just define the policy_namespaces in an array format, like below.

This is similar to PR https://github.com/thoughtbot/administrate/pull/1335 without the backward compatibility of resolve_admin.

```
module Admin
  class ApplicationController < ApplicationController
    include Administrate::Punditize

    def policy_namespaces
      [:admin]
    end
  end
end
```

pablobm commented:
Nice, thank you! I don't have much time now, but I'll review it a bit more thoroughly when I can.

One thing is that ideally I'd like to have some sort of back-compatibility hook where we show a warning to people using the old interface. Do you know how we could achieve this? If you don't, it's ok, I'll get to it. At least this PR gets us moving.

bhtabor commented:
> One thing is that ideally I'd like to have some sort of back-compatibility hook where we show a warning to people using the old interface. Do you know how we could achieve this? If you don't, it's ok, I'll get to it. At least this PR gets us moving.

Just added pundit policy scope `resolve_admin` deprecation. Thank you!

pablobm commented on app/controllers/concerns/administrate/punditize.rb:
```diff
@@ -7,32 +7,58 @@ module Punditize
+
         def scoped_resource
-          policy_scope_admin super
+          namespaced_policy_scope super
```
Another nitpick. This method relies on a private method, while the two below don't. When reading the code, this made me think how this one is special and different from the others, which I don't think it is.

I think that I would remove the private method and inline it here, and this will look this method more similar to its two siblings below, diminishing the cognitive load. Does that make sense?

```suggestion
        namespaced_scope = policy_namespaces << super
        policy_scope!(pundit_user, namespaced_scope)
```

bhtabor replied:
Totally agree. Done!

pablobm reviewed (commented):
One thing I can't find a good place to put: I think we should remove `OrderPolicy#resolve_admin`, for two reasons:

1. It pollutes the specs with warnings.
2. Our example app should avoid deprecated APIs.

The problem is that then this would break the specs you wrote to check for deprecations. I wonder if we can do this differently, perhaps adding `#resolve_admin` in a `before` block with metaprogramming, then remove it in the `after` block. Thoughts? I'm happy with having a much simpler spec for this, instead of looking

Anyway, this is looking pretty good! I want to give it another review though, so make sure I'm not missing anything as this could have large repercussions for many people.

Also I wonder if @nickcharlton or @littleforest would have some time to have a look too :eyes:

bhtabor commented on app/controllers/concerns/administrate/punditize.rb:
```diff
@@ -7,32 +7,58 @@ module Punditize
+
+      def policy_scope!(user, scope)
+        policy_scope_class = Pundit::PolicyFinder.new(scope).scope!
+        return unless policy_scope_class
```
You're correct. I don't think it is necessary. It was a copy/paste from the Pundit [source code](https://github.com/varvet/pundit/blob/5b22078621cf0249d39dc8c8bb0e037cae779fb0/lib/pundit.rb#L121). When the `resolve_admin` support is eventually dropped, we will remove all this code and just call `Pundit.policy_scope!` from `scoped_resource` instead.

bhtabor commented on app/controllers/concerns/administrate/punditize.rb:
```diff
@@ -7,32 +7,58 @@ module Punditize
       included do
         private
 
+        def policy_namespaces
```
Your reservation makes perfect sense! I've actually thought about this. Didn't want to use the singular `policy_namespace` because it has to return an array. I thought users might choose to return `Foo::Bar` if it is singular. I've refactored the code now to use the class method `pundit_policy_namespace` and users should set `Foo::Bar` instead of the array format `[:foo, :bar]`. I think the array is just an implementation detail from [Pundit](https://github.com/varvet/pundit/blob/5b22078621cf0249d39dc8c8bb0e037cae779fb0/lib/pundit/policy_finder.rb#L74) that we don't need to expose.

pablobm replied:
Actually, I prefer having an overridable method like you had originally (but with the singular name). This is more in the spirit of "no DSLs" that we are trying to achieve with Administrate.

As for whether users should give lists of symbols `[:foo, :bar]` or straight modules `Foo::Bar`, I think it should be the list of symbols. When a user uses `Punditize`, they are already commited to Pundit, so there's no need to hide what's really going on. In fact it'll be easier for the user to understand (I think) that the namespace is `[:foo, :bar]` as it is what they would normally use with Pundit.

Does that make sense?

bhtabor replied:
> When a user uses `Punditize`, they are already commited to Pundit, so there's no need to hide what's really going on. In fact it'll be easier for the user to understand (I think) that the namespace is `[:foo, :bar]` as it is what they would normally use with Pundit.

Makes sense.

bhtabor commented:
> One thing I can't find a good place to put: I think we should remove `OrderPolicy#resolve_admin`, for two reasons:
> 
> 1. It pollutes the specs with warnings.
> 2. Our example app should avoid deprecated APIs.

Totally agree. I've addressed those issues. Thank you!

pablobm reviewed (commented):
We are getting there! Once we resolve the details in the comments, I think this can be merged if my colleagues don't have any other comments.

bhtabor commented:
> We are getting there! Once we resolve the open threads, I think this can be merged if my colleagues don't have any other comments.

I've addressed the open issues and rebased to main. Thank you @pablobm!

pablobm commented:
Thank you @bhtabor :slightly_smiling_face:
